src/Microsoft.Health.Dicom.SqlServer/Features/Schema/VersionedCache.cs

```diff
@@ -0,0 +1,69 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Dicom.SqlServer.Exceptions;
+
+namespace Microsoft.Health.Dicom.SqlServer.Features.Schema
+{
+    internal sealed class VersionedCache<T> : IDisposable where T : class, IVersioned
+    {
+        private readonly ISchemaVersionResolver _schemaVersionResolver;
+        private readonly Dictionary<SchemaVersion, T> _entities;
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+        private volatile T _cachedEntity;
```

Comment:
I have to read up more on the use `volatile` for the double-checked lock pattern. Outside of the memory barrier, I am unsure if it's needed

Reply:
It appears it is indeed needed, as `Lazy<T>` itself leverages a volatile write for this purpose